src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// This class provides helper methods for writing bytes using <see cref="TdsStream"/>
+    /// It extends <see cref="TdsBufferAlloc"/> that manages allocations of bytes buffer for better memory management.
+    /// </summary>
+    internal class TdsWriter : TdsBufferAlloc
```

Comment:
Consider making the writer Disposable and also IDisposableAsync.

This can be then disposed when the Dispose() DisposeAsync() of `TdsStream` is called.